Draft for the new backend (Round Two)

CodedOre opened the pull request:
Since I am back working on "Cawbird 2.0", I am working on a new draft on how the backend is build up (mainly because my [first draft](https://github.com/CodedOre/Cawbird/issues/2) was not good...)

This is not complete yet (still a few things missing), but I want to upload it now so that @IBBoard can take a first look (if you have the time and interest).

A few notes at the end:
- I uploaded it as a PR so we can use the "Reviews" for comments.
- The class listed in the documents are intented to be subclassed in API-specific namespaces (such as `TwitterLegacy` for API v1 and `Twitter` for API v2) which then implement the API calls, but no additional properties.

CodedOre committed:
Initial state of backend-design

CodedOre committed:
Add Collection and it's subclasses

IBBoard commented on Readme.md:
```diff
@@ -1,3 +1,75 @@
+
+## API-Flags
+
+The API-Flags define in which scope a functionality is used and where it get's it's information from.
```
Nice use of badges to make it clear what's happening between different APIs.

IBBoard commented on content/class/Animated.md:
```diff
@@ -0,0 +1,80 @@
+
+`Animated` is a subclass of [`Media`](Media.md) which manages a animated picture (gif) from a platform.
+
+> From a technically standpoint `Animated` is just a video without audio, but since it is intended to loop and has different controls it is a separate subclass.
```
Twitter *used* to supply GIFs, which we'd handle differently to video (`Gdk.PixbufAnimation` vs GStreamer) but now they seem to recompress all GIFs as MPEG.

CodedOre replied:
I also want to make `Animated` to return a video file.

This is again a thing coming from "place all important information in the backend and let the frontend just display it", so it would make sense to have a class for "Animated GIF's", despite it is basically the same as `Video`.

IBBoard commented on content/class/Profile.md:
````diff
@@ -0,0 +1,206 @@
+public string created_at { get; }
+```
+
+When the User created his account.
````
Can we use gender-neutral terms for users, please 🙂 You already had "itself" rather than "himself" earlier (although I'd normally use "themself" to make them sound a bit more human!)

CodedOre replied:
Rewrote this to "When this Profile was created."

IBBoard commented on content/class/Animated.md:
````diff
@@ -0,0 +1,80 @@
+![API: Internal](https://img.shields.io/badge/API-Internal-green?style=flat-square)\
+
+```c#
+public async Gtk.MediaFile load_media ()
````
This looks like it's a new GTK4 thing. I think that's going to HUGELY reduce our supportable distros - https://pkgs.org/search/?q=gtk4

CodedOre replied:
Well, given that I am not really fast (I hope I can target October for finishing this), I think it is reasonable to make the jump to the next GTK...

IBBoard commented on content/class/Media.md:
```diff
@@ -0,0 +1,213 @@
+
+> Neither `Media` nor it's subclass stores the full media. Instead the function loads it and hands it over to the Media Viewer, which clears the media from memory after it is closed.
+
+## Signals
```
I think we need progress signals as well, so that we can update load progress percentages etc

IBBoard commented on content/class/Poll.md:
````diff
@@ -0,0 +1,117 @@
+### `flags` {#field_flags}
+
+```c#
+private Gee.HashSet<PollFlag> flags
````
If these are proper enum values, can we not just bitmask them together rather than dragging in libgee and having a hashset of them?

CodedOre replied:
The HashSet was the closest I could find as being similar to a bitmask in Vala (I mean, it also don't have dictionaries to my knowledge)...

IBBoard replied:
Vala supports bitmasking of flags without any additional collections: https://github.com/IBBoard/cawbird/blob/master/src/list/TweetListEntry.vala#L306

Dictionaries are called [HashMaps](https://valadoc.org/gee-0.8/Gee.HashMap.html) in libgee or [HashTable](https://valadoc.org/glib-2.0/GLib.HashTable.html) in plain glibc.

CodedOre replied:
Alright, didn't know of this (but to my defense I could not find it in the documentation...).

Will look into this.

IBBoard replied:
Yeah, it's probably tricky to search for.

And according to the docs that I've just found then there's a better way to do it in pure Vala!

https://wiki.gnome.org/Projects/Vala/Manual/Enumerated%20types%20(Enums)#Flag_types

Our other code is referencing a C enum that's then imported via the vaapi file, and each value is defined by bit-shifting.

IBBoard commented on content/class/Post.md:
```diff
@@ -0,0 +1,316 @@
+| Field | `id_str`  | `id`      | `id`     |
+| Type  | string    | string    | string   |
+
+### `type` {#property_type}
```
Presumably needs the "internal" badge

IBBoard commented on content/class/Post.md:
```diff
@@ -0,0 +1,316 @@
+
+> Twitter urls have to be generated internally, as there is no field for it.
+>
+> So, not really neccessary for Twitter (as it is only `twitter.com`, which can be generated on the fly), but as Mastodon has multiple servers, we need to specify it for it, should it be implemented.
```
Not necessary to have an internal backing store holding the string (unless we're going to call it a lot and benefit from caching it, which seems unlikely). But it does make sense to have it as a consistent property.

IBBoard commented on content/class/Post.md:
````diff
@@ -0,0 +1,316 @@
+![API: Twitter 1](https://img.shields.io/badge/API-Twitter%201-lightgrey?style=flat-square) ![API: Twitter 2](https://img.shields.io/badge/API-Twitter%202-blue?style=flat-square) ![API: Mastodon](https://img.shields.io/badge/API-Mastodon-purple?style=flat-square)
+
+```c#
+public Thread? thread { get; }
````
We may have to be careful to avoid reference loops between threads and posts (e.g. Vala `weak` references)

IBBoard commented on content/class/Post.md:
````diff
@@ -0,0 +1,316 @@
+public Post? referenced_post { get; }
+```
+
+If an post is an repost or quote, this stores the post reposted or quoted.
````
I'm going to have to get used to using non-Twitter terms like "repost" rather than "retweet"!

IBBoard commented on content/class/Post.md:
```diff
@@ -0,0 +1,316 @@
+
+All entities found in the text, to provide additional informations for the text formatting. See [TextEntity](../structure/TextEntity.md).
+
+> This field might be only available in debug builds, as we should not need this field after formatting.
```
We need text entities (and the raw text) when recreating the formatted text - e.g. when changing whether we hide trailing image URLs etc. I don't know how (or if) Mastadon could support that without a lot of very messy HTML manipulation.

IBBoard commented on content/class/Profile.md:
```diff
@@ -0,0 +1,206 @@
+
+The header image of the User (for the profile page), stored as a [`Picture`](Picture.md).
+
+> Mastodon provides the option to use a animated GIF (which we store as a [`Animated`](Animated.md)) as a avatar. For now, we will not consider this option as it makes the code more complex and use always the static variant.
```
This seems to be confusing "header image" (background) with avatar. Can Mastadon do animated headers, or is this comment against the wrong property? Or is it just a copy-paste because the header supports animation as well? (which sounds like a dreadful idea from the turn of the millennium with GeoCities sites!)

CodedOre replied:
Just a copy-paste issue.

IBBoard commented on content/class/User.md:
```diff
@@ -0,0 +1,134 @@
+| Field | `id_str`  | `id`      | `id`     |
+| Type  | string    | string    | string   |
+
+### `name` {#property_name}
```
Can we call this "display_name"? Because I think of name (fixed ASCII handle) and display name (pretty name that people shove emojis in) and it's consistent with your comment.

IBBoard commented on content/class/User.md:
```diff
@@ -0,0 +1,134 @@
+| Field | `name`    | `name`    | `display_name` |
+| Type  | string    | string    | string         |
+
+### `alias` {#property_alias}
```
"Alias" sounds like an additional name rather than their username/handle. Can we call it username so that we don't have clashes with "name"?

IBBoard commented on content/enum/PostFlag.md:
```diff
@@ -0,0 +1,126 @@
+
+This Post has an reply from the active user.
+
+### `HIDDEN_ON_TIMELINES`
```
Is this different to the various "hidden" states that we have at the moment?  https://github.com/IBBoard/cawbird/blob/master/src/CbTweet.h#L32-L47

Does Mastadon let people post messages that are on their timeline (i.e. if you view a profile) but not in the home timeline feeds of anyone who follows them?

CodedOre replied:
The internal hiding states are not included yet (have to put this onto my todo).

Mastodon allows to hide posts from the "public" timeline (among other options). But since we would probably not include the public timeline anyway, we might need this only as an option for posting.

IBBoard commented on content/enum/PostFlag.md:
```diff
@@ -0,0 +1,126 @@
+| Field | `visibility/unlisted` |
+| Type  | string enum           |
+
+### `HIDDEN_UNLESS_FOLLOWER`
```
I'm guessing this is more granular than setting a protected account.

CodedOre replied:
Indeed it is.

IBBoard commented on content/enum/UserFlag.md:
```diff
@@ -0,0 +1,36 @@
+
+Set if the User moderates it's followers.
+
+> On Twitter, this also set's if Tweets are visible to non-followers, however, this does not apply to Mastodon, where visibility of Toots is set for each seperatly.
```
Does Mastadon automatically apply the visibility per new message posted and old messages are still visible? Or does the API set the appropriate values automatically to all messages depending on the account state?

CodedOre replied:
As far as I understand it the `protected` user and a visible post are completely separate on Mastodon.

IBBoard commented on content/structure/PollOption.md:
```diff
@@ -0,0 +1,76 @@
+
+The position of the option in the Poll.
+
+> Probably unnecessary, for example Mastodon does not include it as well.
```
We can probably just use this as a way to get the ordering correct at load time, and then just rely on the ordering of the array.

IBBoard commented on content/structure/TextEntity.md:
```diff
@@ -0,0 +1,60 @@
+
+`TextEntity` stores information about special entities (like mentions or hashtags) that can be found in a text.
+
+> This is only relevant for the Twitter API, as Mastodon provides an HTML-string where we can probably just replace the entity-links with our values.
```
Replacing values may not be entirely trivial. We might be able to intercept the URL handling and just parse the platform's URLs rather than using custom URLs (as we do now) but we'd have to be careful not to intercept the "open in browser" link as well.

IBBoard commented on organization/class/Home.md:
```diff
@@ -0,0 +1,64 @@
+
+Creates `Home` for a given [`Account`](Account.md) and pulls posts.
+
+## Methods
```
All of the other API definitions just have "inherited from" when the functions match. Is there a reason that these are redefined?

CodedOre replied:
I have to ask what do you refer to as "other API definitions" (I guess other documentations), but I "redefined" it to make clear that it is a different implementation from the parent class.

IBBoard replied:
Sorry, I mean "all of the other times in these API documents that a sub-class inherits a function have just had 'inherited from', but this one gets explicitly defined".

I wondered whether it was copy-and-paste that you forgot to change to "inherited" or whether it was for a reason. And apparently it's for a reason because they re-implement rather than just inheriting 🙂

CodedOre replied:
Actually, I looked at it and noticed that `get_posts` is technically just a getter for `post_list`, so I changed that to be inherited...

IBBoard commented on content/class/Media.md:
```diff
@@ -0,0 +1,213 @@
+
+### `media_url` {#property_media_url}
+
+![API: Twitter 1](https://img.shields.io/badge/API-Twitter%201-lightgrey?style=flat-square) ![API: Twitter 2 (Not yet)](https://img.shields.io/badge/API-Twitter%202%20(Not%20yet)-darkred?style=flat-square) ![API: Mastodon](https://img.shields.io/badge/API-Mastodon-purple?style=flat-square)
```
I'm surprised that this (and some other bits) aren't in Twitter v2 yet. But then I've not looked at the API docs in much depth.

CodedOre replied:
You can already get the media url for Pictures over an inofficial argument, but this is not yet included for videos. So, the v2 API is definitely "Work in Progress".

IBBoard reviewed (commented):
Looks like a good structure, and plenty of references. It would be nice if we can easily support Mastadon.

I think we should make generic implementations of the classes where we can (i.e. where there's no divergence) and only have platform-specific ones where necessary. And I'm assuming we'll only have one Twitter implementation that we'll swap to v2 as functions arrive, rather than having both v1 and v2 implementations concurrently.

CodedOre committed:
Minor changes after input.

CodedOre committed:
Change from API-Flags to ScopeAPI-Table

For a better distinction between where we want to use a element and where we get the information for it from.
